### PR TITLE
Issue 295: Fix multiple thread conditional appending

### LINKED ISSUE #295: Correctly handle conditional append failure
**Problem description**
If there are two threads A and B. A appends to the stream and suddenly B appends to the same stream and succeeded. A will receive a ConditionalCheckFailure error in this case. The current impl will assume if error happens to an append, the following appends will also fail, but this is not the case for ConditionalCheckFailure as it only returns error for that specfic append. So there is a race condition that the  ConditionalCheckFailure will never be received by the application

**Problem location**
ByteWriter

**Suggestions for an improvement**

### PR DESCRIPTION
**Pull request by Tristan1900**
**Change log description**  
Fix race condition in multi thread conditional append.
Remove sync API in byte API to make it consistent. 
Remove previous fix #209 since Pravega side issue has been fixed. See issue #198.
Use custom `pravega::error::Error` instead of `std::io::Error` for write result so caller can call match on the error type to check if ConditionalCheckFailure happens.

**Purpose of the change**  
Fix #295 #198

**What the code does**  
Current impl is to only hold the latest EventHandle. The fix is to hold all the EventHandles and checks each one to see if there is a ConditionalCheckFailure.

If ConditionalCheckFailure happens, reactor will set it status to be `need_reset` and reject all new requests since caller may send append before receiving the failure. So caller need to manually call `reset` before continue the append.
**How to verify it**

### CODE REVIEW

File: src/byte/writer.rs [hunk 1]
```diff
@@ -115,17 +116,36 @@ impl Write for ByteWriter {
...
+        let mut index = 0;
+        for handle in self.event_handles.as_mut().unwrap().iter_mut() {
+            if let Ok(res) = handle.try_recv() {
+                res.map_err(|e| Error::new(ErrorKind::Other, format!("{:?}", e)))?;
+                index += 1;
+            } else {
+                break;
+            }
+        }
+        self.event_handles
```
Comment (tkaitchuck) [inline src/byte/writer.rs:134]: Would a single loop with `front` and `pop` be clearer?

File: integration_test/src/byte_reader_writer_tests.rs [hunk 2]
```diff
@@ -231,59 +226,32 @@ fn test_write_and_read_with_workload(writer: &mut ByteWriter, reader: &mut ByteR
...
+    writer1.reset().await.expect("reset writer");
+    let _num = writer1.write(&payload).await.expect("writer1 write payload");
+    assert_eq!(writer1.current_offset(), 3072);
+    writer1.flush().await.expect("writer1 flush");
+    writer1.seek_to_tail().await;
+    assert_eq!(writer1.current_offset(), 3072);
     info!("test byte stream multiple writers concurrent append passed");
 }
-
-async fn test_async_write_and_read(factory: &ClientFactory, stream: ScopedStream) {
```
[the remark this answers is not in the document]
Comment (Tristan1900) [reply integration_test/src/byte_reader_writer_tests.rs:263]: Yeah, the simple_write_and_read was testing sync API and it's testing the async API now.

File: src/byte/reader.rs [hunk 3]
```diff
@@ -69,16 +71,6 @@ pub struct ByteReader {
     factory: ClientFactoryAsync,
 }
 
-impl Read for ByteReader {
```
[the remark this answers is not in the document]
Comment (Tristan1900) [reply src/byte/reader.rs:72]: I bumped the version to 0.3 before when some API changed but 0.3 is not released. So I guess we should still remian in the 0.3?

File: src/byte/writer.rs [hunk 4]
```diff
@@ -167,86 +140,113 @@ impl ByteWriter {
...
-        bytes_to_write
+        self.event_handles.push_back(oneshot_receiver);
+
+        while let Some(handle) = self.event_handles.front_mut() {
+            if let Ok(res) = handle.try_recv() {
+                match res {
+                    Ok(_) => {
+                        self.event_handles.pop_front().expect("remove successful handle");
+                    }
+                    Err(e) => return Err(e),
```
[the remark this answers is not in the document]
Comment (Tristan1900) [reply src/byte/writer.rs:171]: Right!

File: src/byte/writer.rs [hunk 5]
```diff
@@ -167,86 +140,113 @@ impl ByteWriter {
...
-            Ok(())
+    pub async fn flush(&mut self) -> Result<(), Error> {
+        while self.event_handles.front().is_some() {
+            let handle = self.event_handles.pop_front().expect("get first handle");
+            match handle.await {
+                Ok(res) => {
+                    res?;
+                }
+                Err(e) => {
+                    return Err(Error::InternalFailure {
```
[the remark this answers is not in the document]
Comment (Tristan1900) [reply src/byte/writer.rs:199]: Yeah but the `source` in `context` can only be one type. `InternalFailure` could be caused by a few different interal error types. I don't think user need to know those types of error such as oneshot failure and metadataClientError, so I map them to `InternalError`.

[the remark this answers is not in the document]
Comment (tkaitchuck) [reply src/byte/writer.rs:195]: That makes sense. Could the match be an [map_err](https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err)?

**Review by tkaitchuck — APPROVED** [review]

**Pull request merged**